core/matcher.go

```diff
@@ -165,15 +165,30 @@ func isValidHitChatRule(message *models.Message, rule models.Rule, processedInpu
 	if rule.Hear == "" {
 		// Get all the args that the message sender supplied
 		args := utils.RuleArgTokenizer(processedInput)
+		var optionalArgs int
+		var requiredArgs int
+		// take note of all optional args that end with a '?'
+		for _, arg := range rule.Args {
+			if strings.HasSuffix(arg, "?") {
+				optionalArgs++
+			}
+		}
+		// ensure we only require args that don't end with '?'
+		requiredArgs = len(rule.Args) - optionalArgs
 		// Are we expecting a number of args but don't have as many as the rule defines? Send a helpful message
-		if len(rule.Args) > 0 && len(args) < len(rule.Args) {
+		if len(rule.Args) > 0 && requiredArgs > len(args) {
 			msg := fmt.Sprintf("You might be missing an argument or two. This is what I'm looking for\n```%s```", rule.HelpText)
 			message.Output = msg
 			return false
 		}
 		// Go through the supplied args and make them available as variables
 		for i, arg := range rule.Args {
```

Comment:
I think you want to take a closer look at this loop. Currently when you have a rule with:
```yaml
respond: args
args:
  - foo?
  - bar?
```
With an input of `args ok then`, both args are made available as `foo?` and `bar?`. When neither is supplied, ie. `args`, the `?` is stripped correctly from both.

With a rule of:
```yaml
respond: args
args:
  - foo
  - bar?
```
With an input of `args ok`, both are assigned correctly (sans `?`), but with an input of `args ok then`, `bar` is undefined because the `?` is not stripped.

Reply:
Thanks for the feedback, @wass3r. I think this issue stems from me not accounting for the fact that index starts at 0 instead of 1, thus when I looked at `len(args)` I expected it to be +1. What do you think of fixing it via https://github.com/target/flottbot/pull/86/files#diff-5a810f9e7c24fde3e2d5b06e31ebddfdR189?